### PR TITLE
Improve AD function interfaces with bitmasked options

### PR DESCRIPTION
**Pull request by vaithak**
This will extend existing functions to have support for options like vector mode.

A current working example is:
- `clad::differentiate <clad::opts::vector_mode> (fn)` - first-order differentiation of the function `fn` using vector forward mode

Another example which is not yet supported but may be helpful for the future:
- `clad::gradient <clad::opts::use_enzyme, clad::opts::vector_mode> (fn)` - first-order gradient computation (reverse mode AD) using enzyme's vector mode.

### CODE REVIEW

File: include/clad/Differentiator/Differentiator.h [hunk 1]
```diff
@@ -291,9 +292,11 @@ namespace clad {
   /// \param[in] fn function to differentiate 
   /// \param[in] args independent parameter information 
   /// \returns `CladFunction` object to access the corresponding derived function.
-  template <signed N = 1, typename ArgSpec = const char*, typename F,
+  template <unsigned N = 1, unsigned BitMaskedOpts = 0,
+            typename ArgSpec = const char*, typename F,
             typename DerivedFnType = ExtractDerivedFnTraitsForwMode_t<F>,
             typename = typename std::enable_if<
+                !(BitMaskedOpts & clad::opts::vector_mode) &&
```
Comment (vgvassilev) [inline include/clad/Differentiator/Differentiator.h:299]: Can we hide the bit operation behind an interface. Eg something like HasOption or IsEnabled…

Comment (vaithak) [reply include/clad/Differentiator/Differentiator.h]: added the option and corresponding helper in `CladConfig.h` so that it can be accessed from `Differentiator.h` as well as other files.

File: lib/Differentiator/DiffPlanner.cpp [hunk 2]
```diff
@@ -537,51 +538,68 @@ namespace clad {
...
         request.RequestedDerivativeOrder = derivativeOrder;
+
+        unsigned const bitmasked_opts_value =
+            FD->getTemplateSpecializationArgs()
+                ->get(1)
+                .getAsIntegral()
+                .getZExtValue();
+        if (clad::HasOption(bitmasked_opts_value, clad::opts::use_enzyme)) {
+          request.use_enzyme = true;
+        }
```
Comment (vgvassilev) [inline lib/Differentiator/DiffPlanner.cpp:564]:
```suggestion
        if (clad::HasOption(bitmasked_opts_value, clad::opts::use_enzyme))
          request.use_enzyme = true;
```

I believe we could configure clang-format or clang-tidy to find these.

File: test/Enzyme/ForwardMode.C [hunk 3]
```diff
@@ -13,5 +13,5 @@ double f(double x, double y) { return x * y; }
 // CHECK-NEXT:}
 
 int main(){
-     auto f_dx = clad::differentiate<clad::opts::use_enzyme>(f, "x");
+     auto f_dx = clad::differentiate<1, clad::opts::use_enzyme>(f, "x");
```
Comment (vgvassilev) [inline test/Enzyme/ForwardMode.C:16]: That's a bit annoying. I am not sure if we could elide this if we provide something like: `clad::opts::default` as a default argument to `differentiate`...

**Review by parth-07 — APPROVED** [review]
Looks good.

File: .clang-tidy [hunk 4]
```diff
@@ -0,0 +1,32 @@
+Checks: >
```
Comment (parth-07) [inline .clang-tidy:1]: clang tidy changes should not be part of this pull request.

Comment (vaithak) [reply .clang-tidy:1]: yeah, running those to get an idea of what constraints to add in clang tidy, will remove once this is finalised: https://github.com/vgvassilev/clad/pull/584/files

File: include/clad/Differentiator/CladConfig.h [hunk 5]
```diff
@@ -6,6 +6,17 @@
 #include <cstdlib>
 #include <memory>
 
+namespace clad {
+enum opts {
+  use_enzyme = 1 << 0,
+  vector_mode = 1 << 1,
+}; // enum opts
+
+constexpr bool HasOption(const unsigned bitmasked_opts, const unsigned opt) {
```
[the remark this answers is not in the document]
Comment (vgvassilev) [reply include/clad/Differentiator/CladConfig.h:15]: Let's disable this one... we are not that modern :)

File: include/clad/Differentiator/Differentiator.h [hunk 6]
```diff
@@ -327,19 +328,20 @@ namespace clad {
...
-  vector_forward_differentiate(
-      F fn, ArgSpec args = "",
-      DerivedFnType derivedFn = static_cast<DerivedFnType>(nullptr),
-      const char* code = "") {
-    assert(fn && "Must pass in a non-0 argument");
-    return CladFunction<DerivedFnType, ExtractFunctorTraits_t<F>, true>(
-        derivedFn, code);
+      annotate("D")))
+  differentiate(F fn, ArgSpec args = "",
+                DerivedFnType derivedFn = static_cast<DerivedFnType>(nullptr),
```
[the remark this answers is not in the document]
Comment (vgvassilev) [reply include/clad/Differentiator/Differentiator.h:340]: Nope, I'd be in favor of CamelCase...

[the remark this answers is not in the document]
Comment (vaithak) [reply include/clad/Differentiator/Differentiator.h:340]: kept this as `aNy_Case` because there can be cases like taking function decl as parameter, which is many times `FD` or `F`

File: .clang-tidy (hunk 4 above)
Comment (parth-07) [reply .clang-tidy:1]: Oh, okay. Sounds good.

File: test/Enzyme/ForwardMode.C (hunk 3 above)
Comment (vaithak) [reply test/Enzyme/ForwardMode.C:16]: Differentiate has template `<unsigned N=1, unsigned BItMaskedOpts = 0 ....>`, if we want to give non-default value to the second template parameter. I don't see any other way without providing value to the first template parameter.
I think C++-20 allows providing named values to template params, for ex.: `BitMaskedOpts = clad::opts::use_enzyme` but that can be added in the long future.

Comment (parth-07) [reply test/Enzyme/ForwardMode.C:16]: Yes it is slightly annoying, but I don't think we can get away with this if we want to use the same interface, `clad::differentiate`, for both vector and non-vector forward mode.

Comment (vgvassilev) [reply test/Enzyme/ForwardMode.C:16]: Can we somehow mangle the value in the options? Like using the first 3 bits to store the value derivative order.. I don't expect a value greater than 3 to be honest..

Comment (vaithak) [reply test/Enzyme/ForwardMode.C:16]: That is a good point. Will try to make this work and simplify the interface.

Comment (vaithak) [reply test/Enzyme/ForwardMode.C]: added the derivative order in the bitmask options 👍🏼

File: lib/Differentiator/DiffPlanner.cpp [hunk 7]
```diff
@@ -537,51 +538,72 @@ namespace clad {
...
       DiffRequest request{};
 
-      // Check if enzyme is requested in case of Forward Mode or reverse Mode
-      if (A->getAnnotation().equals("D") || A->getAnnotation().equals("G")) {
-        signed opts_value = FD->getTemplateSpecializationArgs()
-                                ->get(0)
-                                .getAsIntegral()
-                                .getZExtValue();
+      if (A->getAnnotation().equals("D")) {
+        request.Mode = DiffMode::forward;
```
[the remark this answers is not in the document]
Comment (vaithak) [reply lib/Differentiator/DiffPlanner.cpp:549]: Should we suppress this type of warning in clang-tidy?

File: lib/Differentiator/DiffPlanner.cpp [hunk 8]
```diff
@@ -537,51 +538,72 @@
...
 
-        if (opts_value == -1) {
+        // bitmask_opts is a template pack of unsigned integers, so we need to
+        // do bitwise or of all the values to get the final value.
+        unsigned bitmasked_opts_value = 0;
+        for (auto const& arg :
+             FD->getTemplateSpecializationArgs()->get(0).pack_elements()) {
```
[the remark this answers is not in the document]
Comment (vaithak) [reply lib/Differentiator/DiffPlanner.cpp:555]: same here?

[the remark this answers is not in the document]
Comment (vgvassilev) [reply lib/Differentiator/DiffPlanner.cpp:555]: Yes, there seems to be a way to ignore pointer conversions only?

[the remark this answers is not in the document]
Comment (vaithak) [reply lib/Differentiator/DiffPlanner.cpp:555]: I don't think so, couldn't find anything here: https://clang.llvm.org/extra/clang-tidy/checks/list.html

[the remark this answers is not in the document]
Comment (vaithak) [reply lib/Differentiator/DiffPlanner.cpp:566]: yup, it worked 👍🏼

File: include/clad/Differentiator/Differentiator.h [hunk 9]
```diff
@@ -483,3 +488,14 @@ namespace clad {
 // Enable clad after the header was included.
 // FIXME: The header inclusion should be made automatic if the pragma is seen.
 #pragma clad ON
+
+// // create a function fn which always return single integer,
```
Comment (vgvassilev) [inline include/clad/Differentiator/Differentiator.h:492]: Why the double comments here?

Comment (vaithak) [reply include/clad/Differentiator/Differentiator.h]: my bad, I was testing something earlier

File: include/clad/Differentiator/CladConfig.h [hunk 10]
```diff
@@ -6,6 +6,37 @@
 #include <cstdlib>
 #include <memory>
 
+namespace clad {
+// The aim is to have a single unsigned integer for storing both, the
+// differentiation order and the options. The differentiation order is
+// stored in the lower 8 bits, and the options are stored in the upper
+// 24 bits. The differentiation order is stored in the lower 8 bits
+// thus allowing for a maximum differentiation order of 2^8 - 1 = 255.
+const unsigned ORDER_BITS = 8;
```
Comment (parth-07) [inline include/clad/Differentiator/CladConfig.h:15]: Minor: this can be constexpr too.

**Comment by vgvassilev** [comment]
Can you squash the second and the last commit into the first and capitalize the clang-tidy commit?

File: include/clad/Differentiator/CladConfig.h [hunk 11]
```diff
@@ -6,6 +6,37 @@
...
+constexpr unsigned GetDerivativeOrder(unsigned const bitmasked_opts) {
+  return bitmasked_opts & ORDER_MASK;
+}
+
+constexpr bool HasOption(unsigned const bitmasked_opts, unsigned const option) {
+  return (bitmasked_opts & option) == option;
+}
+
+constexpr unsigned GetBitmaskedOpts() { return 0; }
+constexpr unsigned GetBitmaskedOpts(unsigned const first) { return first; }
```
Comment (parth-07) [inline include/clad/Differentiator/CladConfig.h:32]: Why do we need this overload?

Comment (vaithak) [reply include/clad/Differentiator/CladConfig.h:32]: This is needed because we unfold this recursively at compile time, as defined in the function below it:
```cpp
template <typename... Opts>
constexpr unsigned GetBitmaskedOpts(unsigned const first, Opts... opts) {
  return first | GetBitmaskedOpts(opts...);
}
```

**Comment by vaithak** [comment]
> Can you squash the second and the last commit into the first and capitalize the clang-tidy commit?

done 👍🏼

File: include/clad/Differentiator/CladConfig.h (hunk 11 above)
Comment (vgvassilev) [reply include/clad/Differentiator/CladConfig.h:32]: I am wondering if we can provide some enum for these magic numbers like `1, 2 ...` to make the code more readable... Maybe up to 3rd order should be reasonable..

Comment (vaithak) [reply include/clad/Differentiator/CladConfig.h:32]: I think it won't be a good experience while implementing because it introduces too many characters into the interface, like `clad::differentiate<clad::opts::order_2>`.  

Whereas, `clad::differentiate<2>` can simply be read as differentiate 2 times.

Comment (vgvassilev) [reply include/clad/Differentiator/CladConfig.h:32]: People would be still able to write it that assuming we get the enum values right.

Comment (vaithak) [reply include/clad/Differentiator/CladConfig.h:32]: Sounds good then, I will add the enums, any preferences for the naming? Or, does `clad::opts::order_1/2/3` work?

Comment (vgvassilev) [reply include/clad/Differentiator/CladConfig.h:32]: Maybe: `clad::opts::order::{first, second, third}`?

I realize as part of this PR we should update the documentation to show the new functionality...

Comment (vaithak) [reply include/clad/Differentiator/CladConfig.h:32]: Sure, will add it 👍🏼

File: docs/userDocs/source/user/UsingEnzymeWithinClad.rst [hunk 12]
```diff
@@ -35,7 +35,7 @@ The following code snippet shows how one can request Clad to generate gradients
     double foo(double* arr) { return arr[0] * arr[1]; }
 
     int main(){
-        auto grad = clad::gradient<clad::opts::use_enzyme>(f1);
+        auto grad = clad::gradient<clad::opts::use_enzyme>(foo);
```
Comment (vgvassilev) [inline docs/userDocs/source/user/UsingEnzymeWithinClad.rst:38]: Let's change the identifier here instead of `foo` to something else. Eg. `sq_array`?

File: docs/userDocs/source/user/UsingVectorMode.rst [hunk 13]
```diff
@@ -0,0 +1,39 @@
+Using Vector Mode for Differentiation
+**************************************
+
+For forward mode AD, the restriction is that the function can be only be differentiated with respect to a single input variable.
```
Comment (vgvassilev) [inline docs/userDocs/source/user/UsingVectorMode.rst:4]: Can you configure your editor to break at 80 columns and trim trailing whitespace?

**Review by vgvassilev — COMMENTED** [review]
Apologies to take so long to merge this PR. I believe that's the last round.

**Comment by vaithak** [comment]
> Apologies to take so long to merge this PR. I believe that's the last round.

No problem, these are user-facing changes so the details matter.

**Review by vgvassilev — APPROVED** [review]
Excellent job!
